### PR TITLE
Separate clang lint?

### PR DESCRIPTION
**Pull request by clee2000**
25 min -> 17 + 13 min, which is still not as fast as I want it to be but I'll take it
Lintrunner provides some parallelism by default, but it's not perfect

Reducing fetch-depth from all to 1 further reduces time by ~2-3 minutes

From non clang's logs:
```
2024-02-09T22:05:39.5297616Z Requirement already satisfied: PyYAML==6.0 in /opt/conda/lib/python3.11/site-packages (6.0)
2024-02-09T22:12:23.6164708Z Collecting black==23.12.1
```
I don't know why this part takes so long, maybe it's just buffering?  Clang version doesn't show this issue

See https://hud.pytorch.org/pytorch/pytorch/commit/5a750c80354d01f7d391f2e817590cc8b587ac9a

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -15,53 +15,32 @@ permissions: read-all
...
-          lint.json || true
-
-        exit $RC
+  lintrunner-noclang:
+    uses: pytorch/test-infra/.github/workflows/linux_job.yml@main
+    with:
+      timeout: 120
+      runner: linux.2xlarge
+      docker-image: pytorch-linux-jammy-cuda11.8-cudnn8-py3.9-linter
+      fetch-depth: 0
```
Comment (huydhn) [inline .github/workflows/lint.yml:38]: Looking at this change and I realize if we need to use `fetch-depth: 0` here for lint job.  I couldn't remember why a full checkout is needed here.  A minute saved is a minute gained?

**Review by huydhn — APPROVED** [review]
This is a nice tweak.  And I thought lintrunner could run well in parallel, but the number shows clear improvement.

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -15,53 +15,32 @@ permissions: read-all
...
-          '"::\(if .severity == "advice" or .severity == "disabled" then "warning" else .severity end) file=\(.path),line=\(.line),col=\(.char),title=\(.code) \(.name)::" + (.description | gsub("\\n"; "%0A"))' \
-          lint.json || true
-
-        exit $RC
+  lintrunner-noclang:
+    uses: pytorch/test-infra/.github/workflows/linux_job.yml@main
+    with:
+      timeout: 120
+      runner: linux.2xlarge
+      docker-image: pytorch-linux-jammy-cuda11.8-cudnn8-py3.9-linter
```
Comment (huydhn) [inline .github/workflows/lint.yml:37]: Back to the question you have earlier about https://github.com/pytorch/pytorch/pull/110502, I think you could also use the smaller linter image `pytorch-linux-focal-linter` here as clang-tidy/format aren't needed

File: .github/workflows/lint.yml (hunk 1 above)
Comment (clee2000) [reply .github/workflows/lint.yml]: Shaves of about 2-3 minutes

**Review by malfet — APPROVED** [review]

File: .github/workflows/lint.yml (hunk 2 above)
Comment (malfet) [reply .github/workflows/lint.yml]: On the other hand it introduces a possibility of nasty splits when one part of the linter is running against packages/settings which are missing in another part. So I would've advised to keep identical image for both, but perhaps cleanup `cuda-11.8` image by deleting the actual binaries (as we are not compiling it against, CUDA just use headers)

File: .github/scripts/lintrunner.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+set -x
+
+# The generic Linux job chooses to use base env, not the one setup by the image
+CONDA_ENV=$(conda env list --json | jq -r ".envs | .[-1]")
+conda activate "${CONDA_ENV}"
+
+CACHE_DIRECTORY="/tmp/.lintbin"
```
Comment (malfet) [inline .github/scripts/lintrunner.sh:8]: Is this directory created by docker build? In that case, woudln't it be missing if other docker image is used for non-clang lints?

**Review by malfet — APPROVED** [review]

File: .github/scripts/lintrunner.sh (hunk 3 above)
Comment (clee2000) [reply .github/scripts/lintrunner.sh:8]: I see .lintbin mentioned in the Docker file for both the cuda and the non cuda builds

**Pull request closed without merging**